Singletons of singletons

Can you single a singleton? Currently the answer is "no", since there's no `Sing` instances for `Sing`s. But should there be? Recently, I found myself wanting something like this when trying to write a `Sing` instance for `Sigma`. As a refresher, here's the definition of `Sigma`:

```haskell
data Sigma (s :: Type) :: (s ~> Type) -> Type where
  (:&:) :: forall s t fst. Sing (fst :: s) -> t @@ fst -> Sigma s t
```

Here's what a `Sing` instance for `Sigma` ought to look like:

```haskell
data instance Sing :: forall s (t :: s ~> Type). Sigma s t -> Type where
  (:%&:) :: forall s t (fst :: s) (sfst :: Sing fst) (snd :: t @@ fst).
            Sing sfst -> Sing snd -> Sing (sfst :&: snd :: Sigma s t)
```

However, notice that we have a field of type `Sing (sfst :: Sing fst)`—a singleton of a singleton! Thus, while you can define a `Sing` instance for `Sigma`, it's currently impossible to use in practice, since you can't reasonably construct things of type `Sing (sfst :: Sing fst)`.

I used to think that in order to make this work, I had to define instances like this:

```haskell
data instance Sing :: forall (b :: Bool). Sing b -> Bool where
  SSFalse :: Sing SFalse
  SSTrue  :: Sing STrue
```

This is unfortunate, since:

1. This would require essentially doubling the number of `Sing` instances in use today. Plus, if we ever needed a singleton of a singleton of a singleton (e.g., `SSSFalse`), then we'd need to _triple_ the number of `Sing` instances in use, and so on.
2. As of GHC 8.6, you can't actually promote data family instance constructors to the type level, so that makes this entire plan impossible.

However, there is another way forward here. You can define a single instance (no pun intended) for all `Sing`s like this:

```haskell
newtype instance Sing :: forall k (a :: k). Sing a -> Type where
  SSing :: forall k (a :: k) (s :: Sing a). Sing a -> Sing s
type SSing = (Sing :: Sing (a :: k) -> Type)
```

That's it! Now you can profitably use the `Sing` instance for `Sigma` like so:

```haskell
type family ProjSigma1 (sig :: Sigma s t) :: s where
  ProjSigma1 ((_ :: Sing fst) :&: _) = fst

projSigma2 :: forall s t (sig :: Sigma s t).
              SingKind (t @@ ProjSigma1 sig)
           => Sing sig -> Demote (t @@ ProjSigma1 sig)
projSigma2 (_ :%&: b) = fromSing b

example :: IO ()
example =
  let f :: Sing (sb :: Sing False) -> Sing (sb :&: True :: Sigma Bool (ConstSym1 Bool))
      f sb = sb :%&: STrue
  in print $ projSigma2 $ f $ SSing SFalse
```

As you can see in that code snippet, one nice advantage of this is that we can define `projSigma2` in essentially the same way that real dependently typed languages do (instead of the [Church-style elimination definition](https://github.com/goldfirere/singletons/blob/72229e75f74a9b0ccb67e156668e57430093f7e1/src/Data/Singletons/Sigma.hs#L50-L61) that we're forced to use currently).

-----

Now that I've extolled the virtues of this idea, it's time to mention the drawbacks. In particular, this whole trick relies on the fact that `Sing` is currently a data family. However, if we were to make `Sing` a type family, as proposed in #318, then we would run into GHC limitations. To see why, consider the `Sing`-as-a-type-family equivalent of the `Sing` instance for `Sing`s:

```haskell
type instance Sing = (Sing :: Sing (a :: k) -> Type)
```

If you compile this, then GHC will give up:

```
$ ~/Software/ghc/inplace/bin/ghc-stage2 --interactive -fprint-explicit-kinds Bug.hs
GHCi, version 8.7.20181007: http://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /home/rgscott/.ghci
[1 of 1] Compiling Bug              ( Bug.hs, interpreted )

Bug.hs:42:15: error:
    • Illegal type synonym family application ‘Sing k’ in instance:
        Sing (Sing k a)
    • In the type instance declaration for ‘Sing’
   |
42 | type instance Sing = (Sing :: Sing (a :: k) -> Type)
   |               ^^^^
```

Blast, foiled by [Trac #12564](https://ghc.haskell.org/trac/ghc/ticket/12564) once again. Unfortunately, I don't know how to work around this, which means that if we switch `Sing` over to become a type family, then this idea is imperiled.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -121,15 +121,31 @@ Changelog for singletons project
   As a result, `singletons` no longer generates instances for `SingI` instances
   for applications of `TyCon{N}` to particular type constructors, as they have
   been superseded by the instances above.
-* Redefine `Σ` such that it is now a partial application of `Sigma`, like so:
+* Changes to `Data.Singletons.Sigma`:
+  * `SSigma`, the singleton type for `Sigma`, is now defined.
+  * New functions `fstSigma`, `sndSigma`, `FstSigma`, `SndSigma`, `currySigma`,
+    and `uncurrySigma` have been added. A `Show` instance for `Sigma` has also
+    been added.
+  * `projSigma1` has been redefined to use continuation-passing style to more
+    closely resemble its cousin `projSigma2`. The new type signature of
+    `projSigma1` is:
 
-  ```haskell
-  type Σ = Sigma
-  ```
+    ```hs
+    projSigma1 :: (forall (fst :: s). Sing fst -> r) -> Sigma s t -> r
+    ```
 
-  One benefit of this change is that one no longer needs defunctionalization
-  symbols in order to partially apply `Σ`. As a result, `ΣSym0`, `ΣSym1`,
-  and `ΣSym2` have been removed.
+    The old type signature of `projSigma1` can be found in the `fstSigma`
+    function.
+  * `Σ` has been redefined such that it is now a partial application of
+    `Sigma`, like so:
+
+    ```haskell
+    type Σ = Sigma
+    ```
+
+    One benefit of this change is that one no longer needs defunctionalization
+    symbols in order to partially apply `Σ`. As a result, `ΣSym0`, `ΣSym1`,
+    and `ΣSym2` have been removed.
 * In line with corresponding changes in `base-4.13`, the `Fail`/`sFail` methods
   of `{P,S}Monad` have been removed in favor of new `{P,S}MonadFail` classes
   introduced in the `Data.Singletons.Prelude.Monad.Fail` module. These classes

--- a/src/Data/Singletons/ShowSing.hs
+++ b/src/Data/Singletons/ShowSing.hs
@@ -268,12 +268,17 @@ ultimately went with.
 -}
 
 ------------------------------------------------------------
--- WrappedSing instance
+-- (S)WrappedSing instances
 ------------------------------------------------------------
 
 instance ShowSing k => Show (WrappedSing (a :: k)) where
-  showsPrec p (WrapSing s) =
-    showParen (p > 10) $ showString "WrapSing " . showsPrec 11 s
+  showsPrec p (WrapSing s) = showParen (p >= 11) $
+    showString "WrapSing {unwrapSing = " . showsPrec 0 s . showChar '}'
+      :: ShowSing' a => ShowS
+
+instance ShowSing k => Show (SWrappedSing (ws :: WrappedSing (a :: k))) where
+  showsPrec p (SWrapSing s) = showParen (p >= 11) $
+    showString "SWrapSing {sUnwrapSing = " . showsPrec 0 s . showChar '}'
       :: ShowSing' a => ShowS
 
 ------------------------------------------------------------

--- a/src/Data/Singletons/Sigma.hs
+++ b/src/Data/Singletons/Sigma.hs
@@ -1,12 +1,18 @@
 {-# LANGUAGE AllowAmbiguousTypes #-}
 {-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
 {-# LANGUAGE GADTs #-}
 {-# LANGUAGE ImpredicativeTypes #-} -- See Note [Impredicative Σ?]
+{-# LANGUAGE MultiParamTypeClasses #-}
 {-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE QuantifiedConstraints #-}
 {-# LANGUAGE RankNTypes #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
 {-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
 
 -----------------------------------------------------------------------------
 -- |
@@ -22,18 +28,34 @@
 ----------------------------------------------------------------------------
 
 module Data.Singletons.Sigma
-    ( Sigma(..), Σ
+    ( -- * The 'Sigma' type
+      Sigma(..), Σ
+    , Sing, SSigma(..), SΣ
+
+      -- * Operations over 'Sigma'
+    , fstSigma, FstSigma, sndSigma, SndSigma
     , projSigma1, projSigma2
     , mapSigma, zipSigma
+    , currySigma, uncurrySigma
+
+      -- * Internal utilities
+      -- $internalutilities
+    , ShowApply,  ShowSingApply
+    , ShowApply', ShowSingApply'
     ) where
 
 import Data.Kind (Type)
 import Data.Singletons.Internal
+import Data.Singletons.ShowSing
 
 -- | A dependent pair.
 data Sigma (s :: Type) :: (s ~> Type) -> Type where
   (:&:) :: forall s t fst. Sing (fst :: s) -> t @@ fst -> Sigma s t
 infixr 4 :&:
+instance (ShowSing s, ShowApply t) => Show (Sigma s t) where
+  showsPrec p ((a :: Sing (fst :: s)) :&: b) = showParen (p >= 5) $
+    showsPrec 5 a . showString " :&: " . showsPrec 5 b
+      :: (ShowSing' fst, ShowApply' t fst) => ShowS
 
 -- | Unicode shorthand for 'Sigma'.
 type Σ = Sigma
@@ -48,20 +70,54 @@ case, and the only reason that GHC currently requires this is due to Trac
 ImpredicativeTypes.
 -}
 
+-- | The singleton type for 'Sigma'.
+data SSigma :: forall s (t :: s ~> Type). Sigma s t -> Type where
+  (:%&:) :: forall s t (fst :: s) (sfst :: Sing fst) (snd :: t @@ fst).
+            Sing ('WrapSing sfst) -> Sing snd -> SSigma (sfst ':&: snd :: Sigma s t)
+infixr 4 :%&:
+
+type instance Sing = SSigma
+instance forall s (t :: s ~> Type) (sig :: Sigma s t).
+         (ShowSing s, ShowSingApply t)
+      => Show (SSigma sig) where
+  showsPrec p ((sa :: Sing ('WrapSing (sfst :: Sing fst))) :%&: (sb :: Sing snd)) =
+    showParen (p >= 5) $
+      showsPrec 5 sa . showString " :&: " . showsPrec 5 sb
+        :: (ShowSing' fst, ShowSingApply' t fst snd) => ShowS
+
+instance forall s t (fst :: s) (a :: Sing fst) (b :: t @@ fst).
+       (SingI fst, SingI b)
+    => SingI (a ':&: b :: Sigma s t) where
+  sing = sing :%&: sing
+
+-- | Unicode shorthand for 'SSigma'.
+type SΣ = SSigma
+
 -- | Project the first element out of a dependent pair.
-projSigma1 :: forall s t. SingKind s => Sigma s t -> Demote s
-projSigma1 (a :&: _) = fromSing a
+fstSigma :: forall s t. SingKind s => Sigma s t -> Demote s
+fstSigma (a :&: _) = fromSing a
+
+-- | Project the first element out of a dependent pair.
+type family FstSigma (sig :: Sigma s t) :: s where
+  FstSigma ((_ :: Sing fst) ':&: _) = fst
 
 -- | Project the second element out of a dependent pair.
---
--- In an ideal setting, the type of 'projSigma2' would be closer to:
---
--- @
--- 'projSigma2' :: 'Sing' (sig :: 'Sigma' s t) -> t @@ ProjSigma1 sig
--- @
---
--- But promoting 'projSigma1' to a type family is not a simple task. Instead,
--- we do the next-best thing, which is to use Church-style elimination.
+sndSigma :: forall s t (sig :: Sigma s t).
+            SingKind (t @@ FstSigma sig)
+         => SSigma sig -> Demote (t @@ FstSigma sig)
+sndSigma (_ :%&: b) = fromSing b
+
+-- | Project the second element out of a dependent pair.
+type family SndSigma (sig :: Sigma s t) :: t @@ FstSigma sig where
+  SndSigma (_ ':&: b) = b
+
+-- | Project the first element out of a dependent pair using
+-- continuation-passing style.
+projSigma1 :: (forall (fst :: s). Sing fst -> r) -> Sigma s t -> r
+projSigma1 f (a :&: _) = f a
+
+-- | Project the second element out of a dependent pair using
+-- continuation-passing style.
 projSigma2 :: forall s t r. (forall (fst :: s). t @@ fst -> r) -> Sigma s t -> r
 projSigma2 f ((_ :: Sing (fst :: s)) :&: b) = f @fst b
 
@@ -76,3 +132,58 @@ zipSigma :: Sing (f :: a ~> b ~> c)
          -> Sigma a p -> Sigma b q -> Sigma c r
 zipSigma f g ((a :: Sing (fstA :: a)) :&: p) ((b :: Sing (fstB :: b)) :&: q) =
   (f @@ a @@ b) :&: (g @fstA @fstB p q)
+
+-- | Convert an uncurried function on 'Sigma' to a curried one.
+--
+-- Together, 'currySigma' and 'uncurrySigma' witness an isomorphism such that
+-- the following identities hold:
+--
+-- @
+-- id1 :: forall a (b :: a ~> Type) (c :: 'Sigma' a b ~> Type).
+--        (forall (p :: Sigma a b). 'SSigma' p -> c @@ p)
+--     -> (forall (p :: Sigma a b). 'SSigma' p -> c @@ p)
+-- id1 f = 'uncurrySigma' @a @b @c ('currySigma' @a @b @c f)
+--
+-- id2 :: forall a (b :: a ~> Type) (c :: 'Sigma' a b ~> Type).
+--        (forall (x :: a) (sx :: Sing x) (y :: b @@ x). Sing ('WrapSing' sx) -> Sing y -> c @@ (sx :&: y))
+--     -> (forall (x :: a) (sx :: Sing x) (y :: b @@ x). Sing ('WrapSing' sx) -> Sing y -> c @@ (sx :&: y))
+-- id2 f = 'currySigma' @a @b @c ('uncurrySigma' @a @b @c f)
+-- @
+currySigma :: forall a (b :: a ~> Type) (c :: Sigma a b ~> Type).
+              (forall (p :: Sigma a b). SSigma p -> c @@ p)
+           -> (forall (x :: a) (sx :: Sing x) (y :: b @@ x).
+                 Sing ('WrapSing sx) -> Sing y -> c @@ (sx ':&: y))
+currySigma f x y = f (x :%&: y)
+
+-- | Convert a curried function on 'Sigma' to an uncurried one.
+--
+-- Together, 'currySigma' and 'uncurrySigma' witness an isomorphism.
+-- (Refer to the documentation for 'currySigma' for more details.)
+uncurrySigma :: forall a (b :: a ~> Type) (c :: Sigma a b ~> Type).
+                (forall (x :: a) (sx :: Sing x) (y :: b @@ x).
+                   Sing ('WrapSing sx) -> Sing y -> c @@ (sx ':&: y))
+             -> (forall (p :: Sigma a b). SSigma p -> c @@ p)
+uncurrySigma f (x :%&: y) = f x y
+
+------------------------------------------------------------
+-- Internal utilities
+------------------------------------------------------------
+
+{- $internal-utilities
+
+See the documentation in "Data.Singletons.ShowSing"—in particular, the
+Haddocks for 'ShowSing' and `ShowSing'`—for an explanation for why these
+classes exist.
+-}
+
+class    (forall (x :: a). ShowApply' f x) => ShowApply (f :: a ~> Type)
+instance (forall (x :: a). ShowApply' f x) => ShowApply (f :: a ~> Type)
+
+class    Show (Apply f x) => ShowApply' (f :: a ~> Type) (x :: a)
+instance Show (Apply f x) => ShowApply' (f :: a ~> Type) (x :: a)
+
+class    (forall (x :: a) (z :: Apply f x). ShowSingApply' f x z) => ShowSingApply (f :: a ~> Type)
+instance (forall (x :: a) (z :: Apply f x). ShowSingApply' f x z) => ShowSingApply (f :: a ~> Type)
+
+class    Show (Sing z) => ShowSingApply' (f :: a ~> Type) (x :: a) (z :: Apply f x)
+instance Show (Sing z) => ShowSingApply' (f :: a ~> Type) (x :: a) (z :: Apply f x)

